pennylane/capture/autograph/ag_primitives.py

```diff
@@ -131,6 +135,157 @@ def assert_iteration_inputs(inputs, symbol_names):
             ) from e
 
 
+def assert_iteration_results(inputs, outputs, symbol_names):
+    """The results of a for loop should have the identical type as the inputs, since they are
+    "passed" as inputs to the next iteration. A mismatch here may indicate that a loop carried
+    variable was initialized with wrong type.
+    """
+
+    for i, (inp, out) in enumerate(zip(inputs, outputs)):
+        inp_t, out_t = jax.api_util.shaped_abstractify(inp), jax.api_util.shaped_abstractify(out)
+        if inp_t.dtype != out_t.dtype or inp_t.shape != out_t.shape:
+            raise AutoGraphError(
+                f"The variable '{symbol_names[i]}' was initialized with the wrong type, or you may "
+                f"be trying to change its type from one iteration to the next. "
+                f"Expected: {out_t}, Got: {inp_t}"
+            )
+
+
+# pylint: disable=too-many-positional-arguments
+def _call_pennylane_for(
+    start,
+    stop,
+    step,
+    body_fn,
+    get_state,
+    set_state,
+    symbol_names,
+    enum_start=None,
+    array_iterable=None,
+):
+    """Dispatch to a PennyLane implementation of for loops."""
+
+    # Ensure iteration arguments are properly initialized. We cannot process uninitialized
+    # loop carried values as we need their type information for tracing.
+    init_iter_args = get_state()
+    assert_iteration_inputs(init_iter_args, symbol_names)
+
+    @qml.for_loop(start, stop, step)
+    def functional_for(i, *iter_args):
+        # Assign tracers to the iteration variables identified by AutoGraph (iter_args in mlir).
+        set_state(iter_args)
+
+        # The iteration index/element (for <...> in) is already handled by the body function, e.g.:
+        #   def body_fn(itr):
+        #     i, x = itr
+        #     ...
+        if enum_start is None and array_iterable is None:
```

Comment:
Just out of curiosity, the builtin `if` block here can be traced properly?

Reply:
This part is based on the structure of the function that is being dealt with (whether is a `for (i, x) in enumerate(array)` or `for x in array`, etc), rather than the inputs, so I believe it doesn't need to be traced. We know at compile time which branch we are on for any given for-loop.